hamilton/function_modifiers/macros.py

```diff
@@ -773,65 +809,152 @@ def final_result(upstream_int: int) -> int:
                 return upstream_int
 
         In all likelihood, you should not be using this, and this is only here in case you want to expose a node for
-        consumption/output later. Setting the namespace in individual nodes as well as in ``pipe`` is not yet supported.
+        consumption/output later. Setting the namespace in individual nodes as well as in ``pipe_input`` is not yet supported.
+
     """
 
     def __init__(
         self,
         *transforms: Applicable,
         namespace: NamespaceType = ...,
+        on_input: base.TargetType = None,
```

Comment:
So I'm not sure that `on_input` makes sense for anything but a single string, and the semantics aren't quite the same. Some thoughts:
1. By default, this applies to the first parameter
2. Does `...` apply to everything?

IMO you should restrict it more -- either a string, or default (which is the first parameter). Also, `@load_from` (which is similar) has the parameter named `target`. But that's confusing, and probably not the best (maybe it should be called `on_input` there?. Thoughts on naming?

Reply:
> So I'm not sure that `on_input` makes sense for anything but a single string, and the semantics aren't quite the same. Some thoughts:
> 
> 1. By default, this applies to the first parameter

Yes, I left the default to chain the first parameter and keep the namespace as is for backwards compatibility 👍 .

> 2. Does `...` apply to everything?

I think we can add `...`, but you could also just use the global `on_input` and pass in a list with the parameters.

> 
> IMO you should restrict it more -- either a string, or default (which is the first parameter). Also, `@load_from` (which is similar) has the parameter named `target`. But that's confusing, and probably not the best (maybe it should be called `on_input` there?. Thoughts on naming?

Ah okay, so the idea I had in mind is to leave flexibility that you can run `pipe_input` on more than one parameter: 3 different cases specifically

1. You can set a global `on_input` and it will apply it to those input arguments, like if we want to run it on the second argument *w-->_add_one --> _add_two --> 2nd arg of foo*
```python
def _add_one(v: int) -> int:
    return v + 1

def _add_two(v: int) -> int:
    return v + 2

@pipe_input(
    step(_add_one).named("a"), 
    step(_add_two).named("b"),
    on_input="w",
    namespace="global",
)
def fool(v: int, w: int) -> int:
    return w
```

2. You can set a local `on_input` and choose which transform runs on which input argument, like  *v-->_add_one --> 1st arg of foo* and *w-->_add_two --> 2nd arg of foo*. 
```python
def _add_one(v: int) -> int:
    return v + 1

def _add_two(v: int) -> int:
    return v + 2

@pipe_input(
    step(_add_one).on_input("v").named("a"), 
    step(_add_two).on_input("w").named("b"),
    namespace="local",
)
def fool(v: int, w: int) -> int:
    return v+w
```

3. You can set have a mixture `on_input` for global and local. The local adds specific transforms on top of what the global has. Here the whole pipeline gets applied to `v`, but only *_add_two* to `w`
```python
def _add_one(v: int) -> int:
    return v + 1

def _add_two(v: int) -> int:
    return v + 2

@pipe_input(
    step(_add_one).named("a"), 
    step(_add_two).on_input("w").named("b"),
    on_input="v"
    namespace="local",
)
def fool(v: int, w: int) -> int:
    return v+w
```

4. The `None` defaults to first parameter and if someone sets `on_input` but not all the `steps` in `pipe_input` get hit, we error out that is unclear what the user wants and he either needs to specify locally or globally or omit.

Reply:
Thanks @jernejfrank  -- this makes a lot of sense.

To sum up the offline conversation we had, we decided to restrict it to a single input (for now), knowing that we can turn it on for more. The reasoning is to reduce supported surface area until we get a request for this feature!